#! /bin/sh /usr/share/dpatch/dpatch-run
## preserve-hidden-parts.dpatch by Ben Hutchings <ben@decadent.org.uk>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: This is a fix for bug #491797.  It changes the DOS partition
## DP: rewriting code to avoid overwriting empty or hidden primary
## DP: partitions so long as their entries aren't reused.  This makes
## DP: fatresize and similar front-ends safe to use in the presence of
## DP: hidden partitions.

@DPATCH@
--- a/libparted/labels/dos.c
+++ b/libparted/labels/dos.c
@@ -1148,13 +1148,18 @@
 	if (!table.mbr_signature)
 		table.mbr_signature = generate_random_id();
 
-	memset (table.partitions, 0, sizeof (DosRawPartition) * 4);
-	table.magic = PED_CPU_TO_LE16 (MSDOS_MAGIC);
+	if (table.magic != PED_CPU_TO_LE16 (MSDOS_MAGIC)) {
+		memset (table.partitions, 0, sizeof (DosRawPartition) * 4);
+		table.magic = PED_CPU_TO_LE16 (MSDOS_MAGIC);
+	}
 
 	for (i=1; i<=4; i++) {
 		part = ped_disk_get_partition (disk, i);
-		if (!part)
+		if (!part) {
+			if (table.partitions [i - 1].type != PARTITION_EMPTY)
+				memset (&table.partitions [i - 1], 0, sizeof (DosRawPartition));
 			continue;
+		}
 
 		if (!fill_raw_part (&table.partitions [i - 1], part, 0))
 			return 0;
